erpnext/utilities/bulk_transaction.py

```diff
@@ -103,6 +103,24 @@ def job(deserialized_data, from_doctype, to_doctype, args):
 
 
 def task(doc_name, from_doctype, to_doctype):
+	hooks = frappe.get_hooks("bulk_transaction_task_mapper")
+	mapper = {}
+	for hook in hooks:
+		mapper.update(frappe.get_attr(hook)())
```

Comment:
This part is being used to get the mapper function from hooks.py

Reply:
@blaggacao I hope I am not wrong. Thanks for your hint.
And it now make more sense to me about Frappe ways.

Reply:
With this, the downstream can do something like this,
https://github.com/kittiu/thai_payroll/pull/7/files

Reply:
PS: your downstream code looks much cleaner, now :+1:

Reply:
> I would advise against dogfooding the mapper within ERPNext itself.
> 
> While this is generally a good idea, consider the downside: now, there's one non-lsp-transparent indirection in the code path, meanig you can't just lsp-jump around (e.g. on GitHub) anymore, so overall the code becomes harder to read and maintain.
> 
> We can avoid this by just putting the hook snippet in place of the orginal code exclusively for eventual downstream use.
> 
> It's also a smaller, more trivial change, which is a plus for review.

You mean not to chance the existing, just provide the hook point?
I just worry that the downstream coder won't know what to do. No sample code.

Reply:
Yeah, got ya. This is true in principle, but I think maintainability primes over education.

We'd probably other tools to educate. For example consistent LSP support for all things. That would serve the same goal, but better.